### ISSUE TITLE
Support for obfuscating attributes in XML

### ISSUE THREAD

**Issue by serranocuesta** [robtimus/obfuscation-xml#2]
Not sure if this is the best place to ask for it but is there any intention to support the obfuscation of attributes in XML.

For example for the following XML:
`<Request User="user" Password="password">Content</Request>`

And an obfuscator created as:
`Obfuscator obfuscator = XMLObfuscator.builder()
        .withElement("password", Obfuscator.fixedLength(0))
        .build();`

Get the following as a result:
`<Request User="user" Password="********">Content</Request>`

Thanks.

**Comment by robtimus** [comment]
Adding support for obfuscating attributes has been on my to do list for a while, but I haven't found a proper way to do so. The XML parser that's used tells me _where_ events like the occurrence of start and end tags occur. That's used to determine which parts of the input to obfuscate. For attributes there is no information on where they start or end. That means that I'd have to manually parse the start element tags, something that's error prone.

I'm thinking about adding a second mode that generates a new document. While this loses the original formatting, it does provide more possibilities. The flow would be like this:
```
Obfuscator obfuscator = XMLObfuscator.builder()
        .reformatting() // working title, I still need to think about it
        .withAttribute("password", Obfuscator.fixedLength(8)) // for any attribute called password
        .withAttribute("password", Obfuscator.fixedLength(8))
                .ofElement("Request")  // for the password attribute of Request elements
        .build();
```

I can't give you a timeline at this moment though, as I have a few other things to solve first.

**Comment by serranocuesta** [comment]
Great.

Thanks for the reply, I'll keep an eye on your repo :)

**Comment by robtimus** [comment]
Let's keep this issue open until I've finished the work.

**Comment by robtimus** [comment]
I wasn't able to support attributes as fully as I wanted to yet - specifying attributes for specific elements is not yet supported. In version 1.2 it is now possible to specify attributes globally though. The following should work for you:

`Obfuscator obfuscator = XMLObfuscator.builder().withAttribute("Password", Obfuscator.fixedLength(0)).build();`

Note that case sensitivity for attributes works the same as for elements - the default is case sensitive, the global setting also holds for attributes, and case sensitivity can be specified per attribute.

**Comment by robtimus** [comment]
Note that the nested obfuscator will lead to this:

`<?xml version='1.0'?><Request User="user" Password="">Content</Request>`

To get your expected output (plus the leading XML declaration), you'd need to use `Obfuscator.all()` instead of `Obfuscator.fixedLength(0)`.

**Comment by robtimus** [comment]
FYI, I've just released version 1.3 that allows attributes to be obfuscated differently for specific elements.

**Comment by serranocuesta** [comment]
Fantastic. Thank you!
